src/hotspot/share/runtime/handshake.cpp

```diff
@@ -565,39 +567,31 @@ HandshakeState::ProcessResult HandshakeState::try_process(HandshakeOperation* ma
 
   Thread* current_thread = Thread::current();
 
-  HandshakeState::ProcessResult pr_ret = HandshakeState::_processed;
-  int executed = 0;
+  HandshakeOperation* op = get_op();
 
-  do {
-    HandshakeOperation* op = pop();
-    if (op != NULL) {
-      assert(SafepointMechanism::local_poll_armed(_handshakee), "Must be");
-      assert(op->_target == NULL || _handshakee == op->_target, "Wrong thread");
-      log_trace(handshake)("Processing handshake " INTPTR_FORMAT " by %s(%s)", p2i(op),
-                           op == match_op ? "handshaker" : "cooperative",
-                           current_thread->is_VM_thread() ? "VM Thread" : "JavaThread");
-
-      if (op == match_op) {
-        pr_ret = HandshakeState::_succeeded;
-      }
+  assert(op != NULL, "Must have an op");
+  assert(SafepointMechanism::local_poll_armed(_handshakee), "Must be");
+  assert(op->_target == NULL || _handshakee == op->_target, "Wrong thread");
 
-      op->prepare(_handshakee, current_thread);
+  log_trace(handshake)("Processing handshake " INTPTR_FORMAT " by %s(%s)", p2i(op),
+                       op == match_op ? "handshaker" : "cooperative",
+                       current_thread->is_VM_thread() ? "VM Thread" : "JavaThread");
 
-      _active_handshaker = current_thread;
-      op->do_handshake(_handshakee);
-      _active_handshaker = NULL;
+  op->prepare(_handshakee, current_thread);
 
-      executed++;
-    }
-  } while (have_non_self_executable_operation());
+  set_active_handshaker(current_thread);
+  op->do_handshake(_handshakee); // acquire, op removed after
```

Comment:
Why acquire? Perhaps:
// peek and handshake with op; op removed after

Reply:
It's mean do_handshake provide acquire thus op is removed after.

Reply:
Ok.